harness/tests/cli/test_auth.py

```diff
@@ -43,9 +43,263 @@ def test_auth_with_store(requests_mock: requests_mock.Mocker, user: Optional[str
+    to make no additional calls (why would a container created by the master have to double-check
+    that it's token in the environment is valid?).  Without Use, some test would be unclear.
+    """
+
+    def __init__(self, token: str) -> None:
+        self.token = token
+
+    def __repr__(self) -> str:
+        return f"Use({self.token!r})"
+
+
+def setenv_optional(key: str, val: Optional[str]) -> None:
+    if val is None:
+        os.environ.pop(key, None)
+    else:
+        os.environ[key] = val
+
+
+class ScenarioSet:
+    """
+    Our login logic is so complex I (rb) can't really keep it all in my head.  So the next best
+    thing is to have exhaustive test cases, both to keep the complex system working and to make sure
+    changes in the system are possible without such a high risk of accidental breakages.
+    """
+
+    def __init__(
+        self,
+        req_user: str,
+        req_pass: str,
+        env_user: str,
+        env_pass: str,
+        env_token: str,
+        cache: str,
+        *results: Any,
+    ) -> None:
+        self.req_user = req_user
+        self.req_pass = req_pass
+        self.env_user = env_user
+        self.env_pass = env_pass
+        self.env_token = env_token
+        self.cache = cache
+        self.results = results
+
+    def __repr__(self) -> str:
+        results = ", ".join(repr(r) for r in self.results)
+        return (
+            f"ScenarioSet({self.req_user!r}, {self.req_pass!r}, {self.env_user!r}, "
+            f"{self.env_pass!r}, {self.env_token!r}, {self.cache!r}, {results})"
+        )
+
+    def scenarios(self) -> Iterator[Tuple[str, str, str, str, str, str]]:
+        for req_user in "yn" if self.req_user == "*" else self.req_user:
+            for req_pass in "yn" if self.req_pass == "*" else self.req_pass:
+                for env_user in "yne" if self.env_user == "*" else self.env_user:
+                    for env_pass in "yn" if self.env_pass == "*" else self.env_pass:
+                        for env_token in "yn" if self.env_token == "*" else self.env_token:
+                            for cache in "ynx" if self.cache == "*" else self.cache:
+                                yield req_user, req_pass, env_user, env_pass, env_token, cache
+
+
+@pytest.mark.parametrize(
+    "scenario_set",
+    [
+        # Explicit user and password, no environment settings, or DET_USER is overridden by
+        # explicitly requested user.
+        #
+        #           requested user (y="user", n=None)
+        #            |   requested password (y="req_pass", n=None)
+        #            |    |   DET_USER env var (y="user", n=unset, e="extra")
+        #            |    |    |    DET_PASS env var (y="env_pass, e=None)
+        #            |    |    |     |   DET_USER_TOKEN (y="env_token", e=None)
+        #            |    |    |     |    |   token cache state (y="cache", n=empty, x="expired")
```

Comment:
Damn, I noticed this towards the end of writing out this table and was in crisis for an hour before I decided that what I had was ok, and I said to myself, "nobody is likely to read this carefully enough to notice the confusion."

lol but you called me out.  I'll try to document it in the code.

I'll also expand a bit more here:

Introducing a different username from `TokenStore.get_active_user()` would cause a whole extra dimension to an already huge table of test cases, but from the perspective of the main body of the `_init_session()` call, which is what I'm interested in testing, it only affects the output of `default_load_user_password` and for that reason all the relevant codepaths are already tested by having a DET_USER with an `e` setting that differs from the explicitly requested user.  I couldn't find a counter example where introducing a variable user from the `get_active_user()` added any code coverage to the code I wanted to test.